### PR TITLE
osmFISH notebook fixes

### PR DESCRIPTION
**Pull request by ambrosejcarr**
@dganguli There are some issues with the notebook that I'm not sure how to solve. I suspect some things got deleted. Can you take a look? 

The major results are still consistent, and the remainder of the stack is passing tests now. I have some minor edits to make on the `LocalMaxPeakFinder` but wanted to verify I hadn't screwed up the notebook before I started changing anything.

### CODE REVIEW

**Comment by ambrosejcarr** [comment]
Ready for review.

**Comment by joshmoore** [comment]
@ambrosejcarr : can you explain why this notebook is loading directly from AWS as opposed to using starfish.data?

**Comment by ambrosejcarr** [comment]
> @ambrosejcarr : can you explain why this notebook is loading directly from AWS as opposed to using starfish.data?

Yes, although it's not a great rationale, so your point is well taken. Much of the data is used for the comparison in addition to the starfish processing, and this is only part of the field of view. I made the minimal fix set to get it to run. 

We're waiting for data from Simone to give us data, at which point we can construct the whole field of view. The right thing might be to hold off on merging the notebook until we have a final version, and only merge the underlying code. 

This notebook does serve as a test, of sorts, for that underlying code, so it was important to surface as a PR.

File: starfish/spots/_detector/trackpy_local_max_peak_finder.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        max_size : float
+            maximum radius-of-gyration of brightness, default None
+        separation : float or tuple
+            Minimum separtion between features. Default is diameter + 1. May be a tuple, see
+            diameter for details.
+        percentile : float
+            Features must have a peak brighter than pixels in this percentile. This helps eliminate
+            spurious peaks.
+        noise_size : float or tuple
+            Width of Gaussian blurring kernel, in pixels Default is 1. May be a tuple, see diameter
```
Comment (ttung) [inline starfish/spots/_detector/trackpy_local_max_peak_finder.py]: spot_diameter?

File: starfish/spots/_detector/trackpy_local_max_peak_finder.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        measurement_type : str ['max', 'mean']
+            name of the function used to calculate the intensity for each identified spot area
+        preprocess : boolean
+            Set to False to turn off bandpass preprocessing.
+        max_iterations : integer
+            max number of loops to refine the center of mass, default 10
+        is_volume : bool
+            if True, run the algorithm on 3d volumes of the provided stack
+        verbose : bool
+            If True, report the percentage completed (default = False) during processing
```
Comment (ttung) [inline starfish/spots/_detector/trackpy_local_max_peak_finder.py]: move default.

File: starfish/spots/_detector/trackpy_local_max_peak_finder.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            )
+
+        # TODO ambrosejcarr: data should always be at least pseudo-3d, this may not be necessary
+        # TODO ambrosejcarr: this is where max vs. sum vs. mean would be parametrized.
+        # here, total_intensity = sum, intensity = max
+        new_colnames = [
+            'x', 'y', 'total_intensity', 'radius', 'eccentricity', 'intensity', 'raw_mass', 'ep'
+        ]
+        if len(image.shape) == 3:
+            attributes.columns = ['z'] + new_colnames
```
Comment (ttung) [inline starfish/spots/_detector/trackpy_local_max_peak_finder.py]: `Indices.Z.value`?

**Comment by ambrosejcarr** [comment]
@dganguli , could you go over the changes required to put together the full field of view and update this notebook to use them? Now that simone's uploaded the data, is it just: 

- [ ] put field of view into starfish spec
- [ ] exchange `aws sync` calls with proper loading
- [ ] find the results and test the copy numbers sequentially across slides

I wouldn't mind using this notebook as an example to hash out #723 (cc @ttung) but in the mean time I could go through the above checklist. 

Let me know if I need to add anything else.

**Comment by dganguli** [comment]
@ambrosejcarr yes those are the remaining items of work. except for "find the results and test the copy numbers sequentially across slides"

i think, for the data he sent, 1 FOV should contain ~33 genes (across all rounds and channels). as such, as we did for other assays, we can generate a copy number comparison scatter plot. x-axis is copy number for pysmFISH, y-axis is copy number for starfish. each point is a gene.

**Comment by dganguli** [comment]
+1 to using this to hash out #723 with @ttung !

**Comment by ambrosejcarr** [comment]
Ready for re-review!

File: MANIFEST.in [hunk 4]
```diff
@@ -1,3 +1,4 @@
 include REQUIREMENTS.txt
 recursive-include sptx_format/schema/ *.json
 recursive-include sptx_format/examples/ *.json
+recursive-include notebooks/data/ *.pkl
```
Comment (ttung) [inline MANIFEST.in]: Is it scalable that we always check in the results?  Also, does it need to be included in the pip pkg?  Looks like you're using a relative path to reference them anyway...

File: notebooks/py/osmFISH.py [hunk 5]
```diff
@@ -3,33 +3,190 @@
...
+from starfish.types import Indices
+
+# EPY: ESCAPE %matplotlib inline
+# EPY: ESCAPE %load_ext autoreload
+# EPY: ESCAPE %autoreload 2
+# EPY: END code
+
+# EPY: START code
+experiment = starfish.data.osmFISH(use_test_data=True)
+stack = experiment['fov_000']['primary']
```
Comment (ttung) [inline notebooks/py/osmFISH.py]: would prefer FieldOfView.PRIMARY_IMAGES

File: notebooks/py/osmFISH.py [hunk 6]
```diff
@@ -3,33 +3,190 @@
...
+# EPY: START code
+plt.figure(figsize=(10, 10))
+plt.imshow(
+    array_for_visualization, 
+    cmap='gray', 
+    vmin=np.percentile(array_for_visualization, 98), 
+    vmax=np.percentile(array_for_visualization, 99.9),
+)
+plt.title('Filtered max projection')
+plt.axis('off');
```
Comment (ttung) [inline notebooks/py/osmFISH.py]:
```suggestion
plt.axis('off')
```

File: notebooks/py/osmFISH.py [hunk 7]
```diff
@@ -3,33 +3,190 @@
...
 
 # EPY: START code
-image = experiment.fov()[FieldOfView.PRIMARY_IMAGES]
-image.show_stack_napari({Indices.CH: 0})
+aldoc_spot_intensities = spot_intensities.sel({Indices.ROUND.value: 0, Indices.CH.value: 0})
+
+plt.hist(aldoc_spot_intensities, bins=20)
+plt.yscale('log')
+plt.xlabel('Intensity')
+plt.ylabel('Number of spots');
```
Comment (ttung) [inline notebooks/py/osmFISH.py]:
```suggestion
plt.ylabel('Number of spots')
```

File: notebooks/py/osmFISH.py [hunk 8]
```diff
@@ -3,33 +3,190 @@
...
+# EPY: START code
+plt.figure(figsize=(10,10))
+plt.imshow(
+    maximum_projection_2d, 
+    cmap = 'gray', 
+    vmin=np.percentile(maximum_projection_2d, 98), 
+    vmax=np.percentile(maximum_projection_2d, 99.9),
+)
+plt.plot(spot_intensities[Indices.X.value], spot_intensities[Indices.Y.value], 'or')
+plt.axis('off');
```
Comment (ttung) [inline notebooks/py/osmFISH.py]:
```suggestion
plt.axis('off')
```

File: notebooks/py/osmFISH.py [hunk 9]
```diff
@@ -3,33 +3,190 @@
...
+plt.plot(benchmark_peaks.x, -benchmark_peaks.y, 'o')
+plt.plot(spot_intensities[Indices.X.value], -spot_intensities[Indices.Y.value], 'x')
+
+plt.legend(['Benchmark: {} spots'.format(benchmark_spot_count),
+            'Starfish: {} spots'.format(starfish_spot_count)])
+plt.title('Starfish x osmFISH Benchmark Comparison');
+# EPY: END code
+
+# EPY: START code
+print("Starfish finds {} fewer spots".format(benchmark_spot_count - starfish_spot_count))
```
Comment (ttung) [inline notebooks/py/osmFISH.py]: Can we assert on this value?

File: starfish/data/__init__.py [hunk 10]
```diff
@@ -53,7 +53,9 @@ def osmFISH(use_test_data: bool=False):
     """
     if use_test_data:
         return Experiment.from_json(
-            'https://d2nhj9g34unfro.cloudfront.net/20181005/osmFISH/experiment.json')
+            'https://d2nhj9g34unfro.cloudfront.net/browse/formatted/20181217/osmFISH/'
```
Comment (ttung) [inline starfish/data/__init__.py]:
```suggestion
            "https://d2nhj9g34unfro.cloudfront.net/browse/formatted/20181217/osmFISH/"
```

File: starfish/data/__init__.py [hunk 11]
```diff
@@ -53,7 +53,9 @@ def osmFISH(use_test_data: bool=False):
...
+            'experiment.json'
```
Comment (ttung) [inline starfish/data/__init__.py]:
```suggestion
            "experiment.json"
```

File: notebooks/py/osmFISH.py [hunk 12]
```diff
@@ -3,33 +3,190 @@
...
+
+# EPY: START code
+benchmark_spot_count = len(benchmark_peaks)
+starfish_spot_count = len(spot_intensities)
+
+plt.figure(figsize=(10,10))
+plt.plot(benchmark_peaks.x, -benchmark_peaks.y, 'o')
+plt.plot(spot_intensities[Indices.X.value], -spot_intensities[Indices.Y.value], 'x')
+
+plt.legend(['Benchmark: {} spots'.format(benchmark_spot_count),
```
Comment (ttung) [inline notebooks/py/osmFISH.py]: no f-style string?

File: notebooks/py/osmFISH.py (hunk 6 above)
Comment (ambrosejcarr) [reply notebooks/py/osmFISH.py]: Without a last-line `;`, irrelevant information about the axes is printed to the output by matplotlib.

File: starfish/data/__init__.py (hunk 10 above)
Comment (ambrosejcarr) [reply starfish/data/__init__.py]: I changed all of `starfish.data`, which was using `'`.

File: notebooks/py/osmFISH.py (hunk 9 above)
Comment (ambrosejcarr) [reply notebooks/py/osmFISH.py]: Yea, good idea. Added.

File: notebooks/py/osmFISH.py (hunk 7 above)
Comment (ambrosejcarr) [reply notebooks/py/osmFISH.py]: see above.

File: notebooks/py/osmFISH.py (hunk 8 above)
Comment (ambrosejcarr) [reply notebooks/py/osmFISH.py]: see above.

File: notebooks/py/osmFISH.py (hunk 6 above)
Comment (ambrosejcarr) [reply notebooks/py/osmFISH.py]: If you would prefer, the alternative is `_ = plt.axis("off")`.

**Pull request merged**